Documentation/proposals/20250122-blackbox-crd.md

```diff
@@ -0,0 +1,153 @@
+
+## Non-Goals
+
+- This propose don't aim to replace the probe crd
+
+## How
+
+We propose creating two new Custom Resource Definitions (CRDs):
+- The BlackboxExporter CRD, which functions similarly to the Prometheus CRD, will manage the deployment configuration and lifecycle of blackbox-exporter instances
+- The BlackboxExporterModule CRD will define module specifications and automatically generate blackbox configurations, analogous to how Alertmanager works in Prometheus Operator
+
+## Alternatives
+Users may need to update or create a Probe CRD to monitor an HTTP endpoint with specific headers or body content. This often requires updating the Blackbox Exporter configuration, such as creating a new module or modifying an existing one. While users can manage this manually, it is a more cumbersome and error-prone process.
+
+## Action Plan
+
+1. Create the BlackboxExporter CRD which will manage blackbox-exporter instances as either a DaemonSet or Deployment
+2. Implement a configuration reloader sidecar container to automatically reload blackbox-exporter when configuration changes, similar to how Prometheus Operator handles configuration updates
+
+## CRDs Draft
+### 1. BlackBoxExporter CRD
+```yaml
+apiVersion: monitoring.zhipu.ai/v1alpha1
+kind: BlackboxExporter
+metadata:
+  name: blackbox-exporter-draft
+spec:
+  # Deployment mode - how the BlackboxExporter should be deployed 
+    # Or Daemonset mode
+  deployMode: Deployment
+  replicas: 2
+  
+  # Key configuration options
+  image: prom/blackbox-exporter
+  timeoutOffset: "0.5"
+  logLevelProber: "info"
+  
+  # Module selection - allows selecting which probe modules to include
+  moduleNamespaceSelector:
+    matchLabels:
+        kubernetes.io/metadata.name: monitoring
+  moduleSelector:
+    matchLabels:
+      type: http-probe
+  
+  # Resource requirements
+  resources:
+    limits:
+      memory: 200Mi
+      cpu: 100m
+    requests:
+      memory: 100Mi
+      cpu: 50m
+      
+  # Advanced configuration options
+  securityContext:
+    runAsNonRoot: true
+    runAsUser: 1000
+  
+  # Network configuration
+  webListenAddresses:
+    - ":9115"
+  externalURL: "https://blackbox.example.com"
+  
+status:
+  # Status fields will be managed by the controller
+  replicas: 2
+  availableReplicas: 2
+  updatedReplicas: 2
+  unavailableReplicas: 0
+  conditions:
+    - type: Available
+      status: "True"
+      lastTransitionTime: "2025-02-28T12:00:00Z"
+      reason: MinimumReplicasAvailable
+      message: "All replicas are available"
+
+```
+### 2. BlackBoxExporterModule CRD
+```
+---
+apiVersion: monitoring.zhipu.ai/v1alpha1
+kind: BlackboxExporterModule
+metadata:
+  name: example-http-probe
+  labels:
+    type: http-probe
+spec:
+  # HTTP probe configuration
+  # Common probe settings
+  timeout: 5s
+  prober: http
+  http:
+    # Request definition
+    method: POST
+    headers:
+      Accept: "application/json"
```

Comment:
see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
maps aren't recommended in Kubernetes APIs.